packages/patternfly-4/react-core/src/components/WithOUIA.tsx

```diff
@@ -0,0 +1,40 @@
+import * as React from 'react';
+import { isOUIAEnvironment, getUniqueId as getOUIAUniqueId } from '../helpers/ouia';
```

Comment:
an absolutely critical note - the ouia id should identify an actual object via its uuid or pkey - having a incremented global as data source there is a massive problem

Reply:
@jschuler @redallen @karelhala ^

Reply:
@RonnyPfannschmidt Why is it a massive problem? It provides unique IDs every page load per-instance of the component that are consistent every page reload. If we were to generate a UUID instead with something like [this solution](https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript), how could we make sure they're unique without a global data source that keeps a record of them all, is inconsistent between page loads, and has the same problem?

Reply:
@redallen i believe you completely misunderstood the use for those ids - they are for identifying application object across potentially different pages, different views and different collections - you cannot generate them randomly or from a counter, you have to take database ids or primary keys

Reply:
we are now iterating on the ouia spec as we noted that the component id specification is lacking a description of the intent to make this absolutely clear

Reply:
to nail this down - after rereading the letter of the spec it is abundantly clear how easy it was to miss the spirit of the spec

Reply:
@RonnyPfannschmidt We are a component library, have no database IDs or primary keys to map to components we create on the page. How do we comply to the spec?

Reply:
@redallen thats why it is critical, a user can pass it in, the user will have the database keys,

for documentation pages/examples, exemplary ids are suitable

Reply:
what's the format of these? haven't seen the spec yet. @redallen maybe something that can be passed in as a json string into another local var?

Reply:
the id should be pretty much expressable as pain string

possible examples are `00000000-0000-0000-0000-000000000000`, `flammable`,  `state-deffered`, `john@example.com`

Reply:
I wasn't sure about the auto generated IDs at first and given the spec we can't really generate them. If I understand correctly the spec these IDs are kinda like helpers to identify what is being rendered. So for instance if you have a list of cards on screen each card would have `data-ouia-component-type="card"` set, but the `data-ouia-component-id` would be probably UUID from database, or whatever to identify which data are actually rendered there.

I think that best way here would be to add `uoiaId` prop to every component that is marked by ouia spec. Consumers can then pass this prop to component and if ouia is enabled we will add it to root element. To ilustrate this for set of cards I will have a list of UUIDs from server and render user data in them
```JSX
import React from 'react';
import { Card, CardBody } from '@patternfly/react-core';

const data = [{
  uuid: '00000000-0000-0000-0000-000000000001',
  name: 'Foo'
}, {
  uuid: '00000000-0000-0000-0000-000000000002',
  name: 'Bar'
}];

export default () => (<React.Fragment>
  {data.map((user) => <Card key={user.id} ouiaId={user.id}>
    <CardBody>{user.name}</CardBody>
  </Card>
)}
</React.Fragment>)
```

The resulting HTML would look like
```HTML
<div class="pf-c-card" data-ouia-component-id="00000000-0000-0000-0000-000000000001" data-ouia-component-type="Card">
  <div class="pf-c-card__body">
    Foo
  </div>
</div>
<div class="pf-c-card" data-ouia-component-id="00000000-0000-0000-0000-000000000002" data-ouia-component-type="Card">
  <div class="pf-c-card__body">
    Bar
  </div>
</div>
```